dxr/vcs.py

```diff
@@ -0,0 +1,327 @@
+        recomb[2] = '/' + recomb[2].lstrip('/') # strip all leading '/', add one back
+        if not upstream.path.endswith('/'):
+            recomb[2] += '/' # Make sure we have a '/' on the end
+        recomb[3] = recomb[4] = recomb[5] = '' # Just those three
+        return urlparse.urlunparse(recomb)
+
+    def find_previous_revisions(self, client):
+        """Find the last revision in which each file changed, for diff links.
+
+        Return a mapping {path: last commit nodes in which file at path changed}
+
+        """
+        last_change = {}
+        for line in client.rawcommand(['previous-revisions']).splitlines():
+            path, node = line.split(':')
+            last_change[path] = node
+        return last_change
+
+    @classmethod
+    def claim_vcs_source(cls, path, dirs, tree):
+        if '.hg' in dirs:
+            dirs.remove('.hg')
+            return cls(path)
+        return None
+
+    def display_rev(self, path):
+        return self.revision[:12]
+
+    def is_tracked(self, path):
+        return path in self.tracked_files
+
+    def generate_raw(self, path):
+        return self.upstream + 'raw-file/' + self.revision + '/' + path
+
+    def generate_diff(self, path):
+        # We generate link to diff with the last revision in which the file changed.
+        return self.upstream + 'diff/' + self.previous_revisions[path] + '/' + path
+
+    def generate_blame(self, path):
+        return self.upstream + 'annotate/' + self.revision + '/' + path
+
+    def generate_log(self, path):
+        return self.upstream + 'filelog/' + self.revision + '/' + path
+
+    def get_contents(self, path, revision):
+        return self.invoke_vcs(['cat', '-r', revision, path])
+
+class Git(Vcs):
+    def __init__(self, root):
+        super(Git, self).__init__(root, 'git')
+        self.tracked_files = set(line for line in
+                                 self.invoke_vcs(['ls-files']).splitlines())
+        self.revision = self.invoke_vcs(['rev-parse', 'HEAD']).strip()
+        self.upstream = self._construct_upstream_url()
+
+    def _construct_upstream_url(self):
+        source_urls = self.invoke_vcs(['remote', '-v']).split('\n')
+        for src_url in source_urls:
+            name, repo, _ = src_url.split()
+            # TODO: Why do we assume origin is upstream?
+            if name == 'origin':
+                if repo.startswith("git@github.com:"):
+                    return "https://github.com/" + repo[len("git@github.com:"):]
+                elif repo.startswith(("git://github.com/", "https://github.com/")):
+                    if repo.endswith(".git"):
+                        repo = repo[:-len(".git")]
+                    if repo.startswith("git:"):
+                        repo = "https" + repo[len("git"):]
+                    return repo
+                warn("Your git remote is not supported yet. Please use a "
+                     "GitHub remote for now, or disable the omniglot "
```

Comment:
The 2nd remediation option doesn't apply anymore. :-) We can just say "Please use a GitHub remote if you would like version control navigation links to show."